File: output/c/Easy/284-number-of-days-between-two-dates.c
/**https://leetcode.com/problems/number-of-days-between-two-dates/ */
//Write a program to count the number of days between two dates.
//The two dates are given as strings, their format is YYYY-MM-DD as shown in the examples.
// 
//Example 1:
//Input: date1 = "2019-06-29", date2 = "2019-06-30"
//Output: 1
//Example 2:
//Input: date1 = "2020-01-15", date2 = "2019-12-31"
//Output: 15
// 
//Constraints:
//	The given dates are valid dates between the years 1971 and 2100.
int daysBetweenDates(char * date1, char * date2){  
    int days = 0;
    int year1 = 0, month1 = 0, day1 = 0;
    int year2 = 0, month2 = 0, day2 = 0;
    sscanf(date1, "%d-%d-%d", &year1, &month1, &day1);
    sscanf(date2, "%d-%d-%d", &year2, &month2, &day2);
    
    int daysOfMonths[12] = {31,28,31,30,31,30,31,31,30,31,30,31};
    if((year1 % 4 == 0 && year1 % 100 != 0) || year1 % 400 == 0)
        daysOfMonths[1] = 29;
    int i = 0;
    for(i = 0; i < month1 - 1; i++)
        days += daysOfMonths[i];
    days += day1;
    
    if((year2 % 4 == 0 && year2 % 100 != 0) || year2 % 400 == 0)
        daysOfMonths[1] = 29;
    else
        daysOfMonths[1] = 28;
    for(i = 0; i < month2 - 1; i++)
        days -= daysOfMonths[i];
    days -= day2;
    
    if(year1 == year2)
        return abs(days);
    else{
        int year = 0;
        if(year1 < year2){
            for(year = year1; year < year2; year++){
                if((year % 4 == 0 && year % 100 != 0) || year % 400 == 0)
                    days += 366;
                else
                    days += 365;
            }
        }
        else{
            for(year = year2; year < year1; year++){
                if((year % 4 == 0 && year % 100 != 0) || year % 400 == 0)
                    days -= 366;
                else
                    days -= 365;
            }
        }
    }
    return abs(days);
}

File: output/java/Easy/284-number-of-days-between-two-dates.java
/**https://leetcode.com/problems/number-of-days-between-two-dates/ */
//Write a program to count the number of days between two dates.
//The two dates are given as strings, their format is YYYY-MM-DD as shown in the examples.
// 
//Example 1:
//Input: date1 = "2019-06-29", date2 = "2019-06-30"
//Output: 1
//Example 2:
//Input: date1 = "2020-01-15", date2 = "2019-12-31"
//Output: 15
// 
//Constraints:
//	The given dates are valid dates between the years 1971 and 2100.
class Solution {
    public int daysBetweenDates(String date1, String date2) {
        int[] days = new int[]{31,28,31,30,31,30,31,31,30,31,30,31};
        int[] daysLeap = new int[]{31,29,31,30,31,30,31,31,30,31,30,31};
        int res = 0;
        String[] date1Array = date1.split("-");
        String[] date2Array = date2.split("-");
        int year1 = Integer.valueOf(date1Array[0]);
        int year2 = Integer.valueOf(date2Array[0]);
        int month1 = Integer.valueOf(date1Array[1]);
        int month2 = Integer.valueOf(date2Array[1]);
        int day1 = Integer.valueOf(date1Array[2]);
        int day2 = Integer.valueOf(date2Array[2]);
        for(int i = year1; i < year2; i++){
            if(isLeap(i)){
                res += 366;
            }else{
                res += 365;
            }
        }
        if(isLeap(year1)){
            for(int i = 0; i < month1 - 1; i++){
                res -= daysLeap[i];
            }
            res -= day1;
        }else{
            for(int i = 0; i < month1 - 1; i++){
                res -= days[i];
            }
            res -= day1;
        }
        if(isLeap(year2)){
            for(int i = 0; i < month2 - 1; i++){
                res += daysLeap[i];
            }
            res += day2;
        }else{
            for(int i = 0; i < month2 - 1; i++){
                res += days[i];
            }
            res += day2;
        }
        return Math.abs(res);
    }
    private boolean isLeap(int year){
        if(year % 400 == 0){
            return true;
        }else if(year % 100 == 0){
            return false;
        }else if(year % 4 == 0){
            return true;
        }else{
            return false;
        }
    }
}